scraper/spiders/thorough_spider.py

```diff
@@ -0,0 +1,78 @@
+# implementation of the thorough spider
+
+import re
+from urllib.parse import urljoin, urlparse
+import scrapy
+from scrapy.linkextractors import IGNORED_EXTENSIONS
+from scraper.items import EmailAddressItem
+
+# scrapy.linkextractors has a good list of binary extensions, only slight tweaks needed
+IGNORED_EXTENSIONS.extend(['ico', 'tgz', 'gz', 'bz2'])
+
+
+def get_extension_ignore_url_params(url):
+    path = urlparse(url).path # conveniently lops off all params leaving just the path
+    extension = re.search('\.([a-zA-Z0-9]+$)', path)
+    if extension is not None:
+        return extension.group(1)
+    else:
+        return "none"  # don't want to return NoneType, it will break comparisons later
+
+
+class ThoroughSpider(scrapy.Spider):
```

Comment:
This is the meat of the implementation, the custom spider which finds urls to follow and email addresses to yield